File: solutions/Golang/minimum_size_subarray_sum_binary_search.go
package main

import (
	"fmt"
	"sort"
)

func minSubArrayLenBinarySearch(target int, nums []int) int {
	prefix := []int{0}
	for _, num := range nums {
		prefix = append(prefix, prefix[len(prefix)-1]+num)
	}

	minLength := int(^uint(0) >> 1)

	for right := 1; right < len(prefix); right++ {
		needed := prefix[right] - target
		leftPos := binarySearchRightmost(prefix, needed, 0, right)
		left := leftPos - 1

		if left >= 0 && left < right {
			if right-left < minLength {
				minLength = right - left
			}
		}
	}

	if minLength == int(^uint(0)>>1) {
		return 0
	}
	return minLength
}

func binarySearchRightmost(arr []int, target int, lo int, hi int) int {
	result := sort.SearchInts(arr[lo:hi], target+1) + lo
	return result
}

func main() {
	nums := []int{2, 3, 1, 2, 4, 3}
	target := 7
	result := minSubArrayLenBinarySearch(target, nums)
	fmt.Println(result)  // 2
}

File: solutions/Java/minimum_size_subarray_sum_binary_search.java
import java.util.Arrays;

public class Main {
    public static int minSubArrayLenBinarySearch(int target, int[] nums) {
        int[] prefix = new int[nums.length + 1];
        prefix[0] = 0;
        for (int i = 0; i < nums.length; i++) {
            prefix[i + 1] = prefix[i] + nums[i];
        }

        int minLength = Integer.MAX_VALUE;

        for (int right = 1; right < prefix.length; right++) {
            int needed = prefix[right] - target;
            int left = binarySearchRightmost(prefix, needed, 0, right) - 1;

            if (left >= 0 && left < right) {
                minLength = Math.min(minLength, right - left);
            }
        }

        return minLength == Integer.MAX_VALUE ? 0 : minLength;
    }

    private static int binarySearchRightmost(int[] prefix, int target, int lo, int hi) {
        while (lo < hi) {
            int mid = lo + (hi - lo) / 2;
            if (prefix[mid] <= target) {
                lo = mid + 1;
            } else {
                hi = mid;
            }
        }
        return lo;
    }

    public static void main(String[] args) {
        int[] nums = {2, 3, 1, 2, 4, 3};
        int target = 7;
        int result = minSubArrayLenBinarySearch(target, nums);
        System.out.println(result);
    }
}
